staging/src/k8s.io/endpointslice/topologies/manager.go

```diff
@@ -0,0 +1,240 @@
+func (m *Manager) activeTopologyForService(serviceKey string, addrType discoveryv1.AddressType) (Topology, bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.activeTopologiesByService[serviceKey]; !ok {
+		m.activeTopologiesByService[serviceKey] = make(map[discoveryv1.AddressType]Topology)
+	}
+
+	topology, ok := m.activeTopologiesByService[serviceKey][addrType]
+	return topology, ok && topology != nil
+}
+
+// setActiveTopologyForService sets the active Topology for the service and
+// addrType.
+//
+// Refer comment under activeTopologyForService() for an understanding of what
+// active Topology means.
+func (m *Manager) setActiveTopologyForService(serviceKey string, addrType discoveryv1.AddressType, topology Topology) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.activeTopologiesByService[serviceKey]; !ok {
+		m.activeTopologiesByService[serviceKey] = make(map[discoveryv1.AddressType]Topology)
+	}
+
+	m.activeTopologiesByService[serviceKey][addrType] = topology
+}
+
+// desiredTopologyForService returns the Topology corresponding to the topology
+// annotation in the service.
+//   - enabled return value signifies that the topology annotation determines an
+//     enabled value.
+//   - exists return value signifies existence of the topology in the topology
+//     manager.
+//
+// The Topology returned represents the "desired state", which could be
+// different from the "current state". Use activeTopologyForService() to get
+// the "current state".
+func (m *Manager) desiredTopologyForService(service *corev1.Service) (topology Topology, enabled, exists bool) {
+	desiredTopologyName, desiredTopologyEnabled := TopologyNameFromAnnotations(service.GetAnnotations())
+	if !desiredTopologyEnabled {
+		return nil, false, false
+	}
+
+	// Special case for backward compatibility: If the topology annotation value
+	// is set to "Auto", we need to return defaultTopology.
+	if desiredTopologyName == "Auto" || desiredTopologyName == "auto" {
+		topology, ok := m.topologiesByName[m.defaultTopologyName]
+		return topology, true, ok
+	}
+
+	for topologyName, topology := range m.topologiesByName {
+		if desiredTopologyName == topologyName || desiredTopologyName == strings.ToLower(topologyName) { //nolint:gocritic // strings.EqualFold() does not exactly match the intent.
+			return topology, true, true
+		}
+	}
+
+	return nil, true, false
+}
+
+// TopologyNameFromAnnotations function returns the value of the topology
+// annotation, if one is present. It also identifies whether the annotation
+// value indicates that the topology is enabled or disabled.
+//
+// The function uses two annotation keys to determine the topology annotation
+// value:
+//   - v1.AnnotationTopologyMode
+//   - v1.DeprecatedAnnotationTopologyAwareHints.
+//
+// If both annotation keys are set, v1.DeprecatedAnnotationTopologyAwareHints
+// has precedence.
+func TopologyNameFromAnnotations(annotations map[string]string) (topologyName string, enabled bool) {
```

Comment:
I know this complicates the code, but I think ideally we'd only support new values with the new annotation (`v1.AnnotationTopologyMode`)